SRC/PORT_IMAGE.CPP

```diff
@@ -0,0 +1,119 @@
+#include <cstring>
+#include <cstdlib>
+#include <SDL_surface.h>
+#include <SDL_image.h>
+#include <ERROR/ERROR.H>
+
+namespace tk_port
+{
+    /**
+     * Replace the extension in the path `name` with `.png`.
+     *
+     * The return value is a static buffer, so the caller does not need
+     * to bother with memory management.  This function is not reentrant.
+     *
+     * @param name Path to (probably) an EFP/FNT name
+     * @return PNG path
+     */
+    char *replace_extension_with_png( const char *name )
+    {
+        static char png_name[80];
+        strncpy( png_name, name, 80 );
+        size_t len = strlen( png_name );
+        if ( len >= 3 )
+        {
+            png_name[len - 3] = 'p';
+            png_name[len - 2] = 'n';
+            png_name[len - 1] = 'g';
+        }
+        return png_name;
+    }
+
+    /**
+     * Load the image data and/or the palette data from a PNG.
+     * @param png_name PNG path.
+     * @param kuva Optional pointer to receive the image data.
+     *             Must be large enough to accommodate it; no bounds checking is or can be done.
+     * @param paletti Optional pointer to a 768-byte char array to receive the palette data.
+     *                The PNG's palette, ranging from 0..255 is rescaled down to 0..63 to match
+     *                VGA hardware and other assumptions in the code.
+     * @param width Optional pointer to hold the width of the image.
+     * @param height Optional pointer to hold the width of the image.
+     * @return True if the load succeeded, False if not.
+     */
+    bool load_png( const char *png_name, char *kuva, char *paletti, int *width, int *height )
+    {
+        SDL_RWops *rwop = SDL_RWFromFile( png_name, "rb" );
+        if ( !rwop )
+        {
+            // assume the file just didn't exist
+            return false;
+        }
+        SDL_Surface *image = IMG_LoadPNG_RW( rwop );
+        if ( !image )
```

Comment:
`SDL_FreeRW( rwop );` before condition to avoid leaking in case of a failure?

Reply:
Well, `error()` is a fatal noreturn, but yeah, good hygiene anyway.